File: 72-edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int len1 = word1.length(); 
        int len2 = word2.length(); 

        int dp[2][len1 + 1]; 

        memset(dp, 0, sizeof dp); 

        for (int i = 0; i <= len1; i++) 
            dp[0][i] = i; 

        for (int i = 1; i <= len2; i++) { 
            for (int j = 0; j <= len1; j++) { 
                if (j == 0) 
                    dp[i % 2][j] = i; 

                else if (word1[j - 1] == word2[i - 1]) { 
                    dp[i % 2][j] = dp[(i - 1) % 2][j - 1]; 
                } 

                else { 
                    dp[i % 2][j] = 1 + min(dp[(i - 1) % 2][j], 
                                           min(dp[i % 2][j - 1], 
                                               dp[(i - 1) % 2][j - 1])); 
                } 
            } 
        } 
        return dp[len2 % 2][len1];
    }
};

File: 72-edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m = len(word1)
        n = len(word2)
        inf = m+n+1
        
        dp = [[inf]*(n+1) for _ in range(m+1)]
        for i in range(m+1):
            dp[i][0] = i
        for j in range(n+1):
            dp[0][j] = j
        dp[0][0] = 0

        for i in range(1, m+1):
            for j in range(1, n+1):
                cost_sub = 1
                if word1[i-1]==word2[j-1]:
                    cost_sub = 0
                
                dp[i][j] = min(
                    cost_sub + dp[i-1][j-1],
                    1 + dp[i-1][j],
                    1 + dp[i][j-1],
                )


        # for i in range(m+1):
        #     for j in range(n+1):
        #         print(dp[i][j], end=' ')
        #     print()

        return dp[m][n]